File: blind_75_leetcode/python/s0572_subtree_of_another_tree.py
import unittest
from typing import Optional


from .s0104_maximum_depth_of_binary_tree import TreeNode, create_tree_from_list


class Solution:
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        def is_identical(t1: Optional[TreeNode], t2: Optional[TreeNode]) -> bool:
            if not t1 and not t2:
                return True
            elif t1 and t2:
                return (
                    t1.val == t2.val
                    and is_identical(t1.left, t2.left)
                    and is_identical(t1.right, t2.right)
                )
            else:
                return False

        if not subRoot:
            return True
        elif not root:
            return False
        else:
            return (
                is_identical(root, subRoot)
                or self.isSubtree(root.left, subRoot)
                or self.isSubtree(root.right, subRoot)
            )


class TestS572(unittest.TestCase):
    def test_works(self):
        self.assertEqual(
            Solution().isSubtree(
                create_tree_from_list([3, 4, 5, 1, 2]), create_tree_from_list([4, 1, 2])
            ),
            True,
        )
        self.assertEqual(
            Solution().isSubtree(
                create_tree_from_list([3, 4, 5, 1, 2, None, None, None, None, 0]),
                create_tree_from_list([4, 1, 2]),
            ),
            True,
        )

File: blind_75_leetcode/src/solution/s0572_subtree_of_another_tree.rs
#![allow(dead_code)]
pub struct Solution {}

use super::s0104_maximum_depth_of_binary_tree::TreeNode;

use std::cell::RefCell;
use std::rc::Rc;
impl Solution {
    pub fn is_subtree(
        root: Option<Rc<RefCell<TreeNode>>>,
        sub_root: Option<Rc<RefCell<TreeNode>>>,
    ) -> bool {
        Solution::is_subtree_ref(&root, &sub_root)
    }

    fn is_subtree_ref(
        root: &Option<Rc<RefCell<TreeNode>>>,
        sub_root: &Option<Rc<RefCell<TreeNode>>>,
    ) -> bool {
        match (root, sub_root) {
            (_, None) => true,
            (None, _) => false,
            _ => {
                Solution::is_identical(root, sub_root)
                    || Solution::is_subtree_ref(&root.as_ref().unwrap().borrow().left, sub_root)
                    || Solution::is_subtree_ref(&root.as_ref().unwrap().borrow().right, sub_root)
            }
        }
    }

    fn is_identical(
        t1: &Option<Rc<RefCell<TreeNode>>>,
        t2: &Option<Rc<RefCell<TreeNode>>>,
    ) -> bool {
        match (t1, t2) {
            (None, None) => true,
            (Some(n1), Some(n2)) => {
                n1.borrow().val == n2.borrow().val
                    && Solution::is_identical(&n1.borrow().left, &n2.borrow().left)
                    && Solution::is_identical(&n1.borrow().right, &n2.borrow().right)
            }
            _ => false,
        }
    }
}

#[cfg(test)]
mod tests_572 {
    use super::*;

    #[test]
    fn it_works() {
        assert_eq!(
            Solution::is_subtree(
                TreeNode::from_vec(vec![Some(3), Some(4), Some(5), Some(1), Some(2)]),
                TreeNode::from_vec(vec![Some(4), Some(1), Some(2)])
            ),
            true
        );
        assert_eq!(
            Solution::is_subtree(
                TreeNode::from_vec(vec![
                    Some(3),
                    Some(4),
                    Some(5),
                    Some(1),
                    Some(2),
                    None,
                    None,
                    None,
                    None,
                    Some(0)
                ]),
                TreeNode::from_vec(vec![Some(4), Some(1), Some(2)])
            ),
            false
        );
    }
}
